File: identity/node/src/documents.ts
import { BaseClient, type FileInput } from './client.js'
import type {
  ExtractDocumentRequest,
  ExtractDocumentResponse,
  ValidateDocumentRequest,
  ValidateDocumentResponse,
} from './types.js'

export class DocumentsResource extends BaseClient {

  /**
   * Extract structured fields from a document image (synchronous).
   *
   * @example
   * const data = await kernaq.documents.extract({
   *   document:     fs.createReadStream('passport.jpg'),
   *   documentType: 'passport',
   *   country:      'GBR',
   * })
   * console.log(data.fields.first_name) // 'John'
   */
  extract(req: ExtractDocumentRequest): Promise<ExtractDocumentResponse> {
    const fields: Record<string, string> = {}
    if (req.documentType) fields['document_type'] = req.documentType
    if (req.country)      fields['country']       = req.country

    return this.upload<ExtractDocumentResponse>('/documents/extract', fields, [
      {
        field:       'document',
        value:       req.document as FileInput,
        filename:    req.documentName ?? 'document.jpg',
        contentType: 'image/jpeg',
      },
    ])
  }

  /**
   * Validate a document — returns validity flags without full OCR (synchronous).
   *
   * @example
   * const result = await kernaq.documents.validate({
   *   document:     fs.createReadStream('id.jpg'),
   *   documentType: 'national_id',
   * })
   * console.log(result.valid) // true
   */
  validate(req: ValidateDocumentRequest): Promise<ValidateDocumentResponse> {
    const fields: Record<string, string> = {}
    if (req.documentType) fields['document_type'] = req.documentType
    if (req.country)      fields['country']       = req.country

    return this.upload<ValidateDocumentResponse>('/documents/validate', fields, [
      {
        field:       'document',
        value:       req.document as FileInput,
        filename:    req.documentName ?? 'document.jpg',
        contentType: 'image/jpeg',
      },
    ])
  }
}

File: identity/go/kernaq/documents.go
package kernaq

import "context"

// DocumentsResource handles stateless document OCR endpoints.
type DocumentsResource struct {
	c *client
}

// Extract extracts structured fields from an ID document. Nothing stored.
func (r *DocumentsResource) Extract(ctx context.Context, req ExtractDocumentRequest) (*ExtractDocumentResponse, error) {
	fields := map[string]string{}
	if req.DocumentType != "" {
		fields["document_type"] = req.DocumentType
	}
	if req.Country != "" {
		fields["country"] = req.Country
	}
	docName := nameOr(req.DocumentName, "document.jpg")
	files := []filePart{{field: "document", reader: req.Document, filename: docName, mime: mimeFromName(docName)}}
	var out ExtractDocumentResponse
	if err := r.c.upload(ctx, "/documents/extract", fields, files, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Validate validates document fields (expiry, required fields present). Nothing stored.
func (r *DocumentsResource) Validate(ctx context.Context, req ValidateDocumentRequest) (*ValidateDocumentResponse, error) {
	fields := map[string]string{}
	if req.DocumentType != "" {
		fields["document_type"] = req.DocumentType
	}
	if req.Country != "" {
		fields["country"] = req.Country
	}
	docName := nameOr(req.DocumentName, "document.jpg")
	files := []filePart{{field: "document", reader: req.Document, filename: docName, mime: mimeFromName(docName)}}
	var out ValidateDocumentResponse
	if err := r.c.upload(ctx, "/documents/validate", fields, files, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
